packages/front-end/components/Metrics/MetricForm.tsx

```diff
@@ -189,44 +189,49 @@ const MetricForm: FC<MetricFormProps> = ({
-      column: current.column || "",
-      inverse: !!current.inverse,
-      ignoreNulls: !!current.ignoreNulls,
-      queryFormat: current.queryFormat || (current.sql ? "sql" : "builder"),
-      cap: current.cap || 0,
-      conversionWindowHours:
-        current.conversionWindowHours || getDefaultConversionWindowHours(),
-      conversionDelayHours: current.conversionDelayHours || 0,
-      sql: current.sql || "",
-      aggregation: current.aggregation || "",
-      conditions: current.conditions || [],
-      userIdTypes: current.userIdTypes || [],
-      userIdColumns: current.userIdColumns || {
-        user_id: current.userIdColumn || "user_id",
-        anonymous_id: current.anonymousIdColumn || "anonymous_id",
-      },
-      timestampColumn: current.timestampColumn || "",
-      tags: current.tags || [],
-      winRisk: (current.winRisk || defaultWinRiskThreshold) * 100,
-      loseRisk: (current.loseRisk || defaultLoseRiskThreshold) * 100,
-      maxPercentChange:
-        (current.maxPercentChange || defaultMaxPercentChange) * 100,
-      minPercentChange:
-        (current.minPercentChange || defaultMinPercentChange) * 100,
-      minSampleSize: current.minSampleSize || defaultMinSampleSize,
+  const transformFormDefaults = () => ({
+    datasource:
+      ("datasource" in current ? current.datasource : datasources[0]?.id) || "",
+    name: current.name || "",
+    description: current.description || "",
+    type: current.type || "binomial",
+    table: current.table || "",
+    denominator: current.denominator || "",
+    column: current.column || "",
+    inverse: !!current.inverse,
+    ignoreNulls: !!current.ignoreNulls,
+    queryFormat: current.queryFormat || (current.sql ? "sql" : "builder"),
+    cap: current.cap || 0,
+    conversionWindowHours:
+      current.conversionWindowHours || getDefaultConversionWindowHours(),
+    conversionDelayHours: current.conversionDelayHours || 0,
+    sql: current.sql || "",
+    aggregation: current.aggregation || "",
+    conditions: current.conditions || [],
+    userIdTypes: current.userIdTypes || [],
+    userIdColumns: current.userIdColumns || {
+      user_id: current.userIdColumn || "user_id",
+      anonymous_id: current.anonymousIdColumn || "anonymous_id",
     },
+    timestampColumn: current.timestampColumn || "",
+    tags: current.tags || [],
+    winRisk: (current.winRisk || defaultWinRiskThreshold) * 100,
+    loseRisk: (current.loseRisk || defaultLoseRiskThreshold) * 100,
+    maxPercentChange:
+      (current.maxPercentChange || metricDefaults.maxPercentageChange) * 100,
+    minPercentChange:
+      (current.minPercentChange || metricDefaults.minPercentageChange) * 100,
+    minSampleSize: current.minSampleSize || metricDefaults.minimumSampleSize,
+  });
+
+  const form = useForm({
+    defaultValues: useMemo(transformFormDefaults, [metricDefaults]),
   });
 
+  useEffect(() => {
+    form.reset(transformFormDefaults());
```

Comment:
Won't this wipe out any progress the user has made so far on the form?  I think it's unlikely that the defaults change while the form is open, but if they do I would probably rather the form just stick with the old default value instead of resetting.

Reply:
If I don't do this, the initial state of the form loads with the hardcoded default values and does not update to the organization settings, making the feature not work. This is because Redux Hook Forms is not reactive, we need to call `form.reset()` explicitly. This is likely because it's a performance-optimized library for forms (compared to Formik or Redux Forms, both of which cause a whole form re-render for every keystroke), so it likely memoizes the values and that's why it needs to be called explicitly.

I am guessing this has to do with race conditions with the render cycles and nested hooks but I'm not entirely sure. Hooks **should** be reactive to dependencies passed into them but it seems like `useForm` is not, but calling `form.reset()` is a workaround for lack of reactiveness of the `useForm` hook.

It looks like the `AuthContext` only fetches the user once on load and doesn't expose any refresh methods. I think it would be safe to leave as is since there is no periodic polling that I can see (at least in the auth and user context providers).

Is there another case I'm missing where the user gets refreshed while they're on the Metrics page?

Reply:
In the SSO branch which I'm going to merge soon, there's logic to refresh the access token if it's about to expire. I believe that might refresh the user too, but I'm not 100% sure. We use the `useSWR` hook pretty extensively throughout the app and that refreshes in the background all the time, like if you blur and refocus the window. I don't think that's used for the user data now, but I wouldn't be surprised if we switch to use that later at some point.

Reply:
Resolving this since the `useMemo` approach fixed it. `form.reset()` is gone. Thanks for clarifying about SSO.